src/test/scala/sigmastate/serialization/DataJsonEncoderSpecification.scala

```diff
@@ -0,0 +1,97 @@
+import special.sigma.AvlTree
+import SType.AnyOps
+import special.collection.Coll
+
+class DataJsonEncoderSpecification extends SerializationSpecification {
+//
+//  def roundtrip[T <: SType](obj: T#WrappedType, tpe: T) = {
+//    val w = SigmaSerializer.startWriter()
+//    DataSerializer.serialize(obj, tpe, w)
+//    val bytes = w.toBytes
+//    val r = SigmaSerializer.startReader(bytes, 0)
+//    val res = DataSerializer.deserialize(tpe, r)
+//    res shouldBe obj
+//    val randomPrefix = arrayGen[Byte].sample.get
+//    val r2 = SigmaSerializer.startReader(randomPrefix ++ bytes, randomPrefix.length)
+//    val res2 = DataSerializer.deserialize(tpe, r2)
+//    res2 shouldBe obj
+//  }
+
+  def testCollection[T <: SType](tpe: T) = {
+    implicit val wWrapped = wrappedTypeGen(tpe)
+    implicit val tT = Evaluation.stypeToRType(tpe)
+    implicit val tagT = tT.classTag
+    implicit val tAny = RType.AnyType
+    forAll { xs: Array[T#WrappedType] =>
+//      println(DataJsonEncoder.encode[SCollection[T]](xs.toColl, SCollection(tpe)))
+//      roundtrip[SCollection[T]](xs.toColl, SCollection(tpe))
+//      roundtrip[SType](xs.toColl.map(x => (x, x)).asWrappedType, SCollection(STuple(tpe, tpe)))
+//
+//      val triples = xs.toColl.map(x => TupleColl(x, x, x)).asWrappedType
+//      roundtrip(triples, SCollection(STuple(tpe, tpe, tpe)))
+//
+//      val quartets = xs.toColl.map(x => TupleColl(x, x, x, x)).asWrappedType
+//      roundtrip(quartets, SCollection(STuple(tpe, tpe, tpe, tpe)))
+//
+//      val nested = xs.toColl.map(x => Colls.fromItems[T#WrappedType](x, x))
+//      roundtrip[SCollection[SCollection[T]]](nested, SCollection(SCollection(tpe)))
+//
+//      roundtrip[SType](
+//        xs.toColl.map { x =>
+//          val arr = Colls.fromItems[T#WrappedType](x, x)
+//          (arr, arr)
+//        }.asWrappedType,
+//        SCollection(STuple(SCollection(tpe), SCollection(tpe)))
+//      )
+    }
+  }
+//
+//  def testTuples[T <: SType](tpe: T) = {
+//    implicit val wWrapped = wrappedTypeGen(tpe)
+//    implicit val tag = tpe.classTag[T#WrappedType]
+//    implicit val tAny = RType.AnyType
+//    forAll { in: (T#WrappedType, T#WrappedType) =>
+//      val (x,y) = (in._1, in._2)
+//      roundtrip[SType]((x, y).asWrappedType, STuple(tpe, tpe))
+//      roundtrip[SType](TupleColl(x, y, x).asWrappedType, STuple(tpe, tpe, tpe))
+//      roundtrip[SType](TupleColl(x, y, x, y).asWrappedType, STuple(tpe, tpe, tpe, tpe))
+//      roundtrip[STuple](Colls.fromItems[Any](x, y, (x, y)), STuple(tpe, tpe, STuple(tpe, tpe)))
+//      roundtrip[STuple](Colls.fromItems[Any](x, y, TupleColl(x, y, x)), STuple(tpe, tpe, STuple(tpe, tpe, tpe)))
+//      roundtrip[STuple](Colls.fromItems[Any](x, y, TupleColl(x, y, (x, y))), STuple(tpe, tpe, STuple(tpe, tpe, STuple(tpe, tpe))))
+//    }
+//  }
+
+  property("Data serialization round trip") {
+
+
+//    forAll { x: Byte => roundtrip[SByte.type](x, SByte) }
```

Comment:
Seems natural to start from simple types first and create Encoder/Decoder pair for them.
Then generalise from it.